pkg/scope/health/health.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package health
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	apps "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/oam-kubernetes-runtime/apis/core/v1alpha2"
+
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/pkg/errors"
+)
+
+const (
+	errNoWorkload            = "could not retrieve workload %q"
+	errNoWorkloadResources   = "could not retrieve resources for workload %q"
+	errResourceNotFound      = "could not retrieve resource %q %q %q"
+	errDeploymentUnavailable = "no ready instance found in %q %q %q"
+
+	defaultTimeout = 10 * time.Second
+)
+
+// UpdateHealthStatus updates the status of the healthscope based on workload resources.
+func UpdateHealthStatus(ctx context.Context, client client.Client, healthScope *v1alpha2.HealthScope) error {
+	timeout := defaultTimeout
+	if healthScope.Spec.ProbeTimeout != nil {
+		timeout = time.Duration(*healthScope.Spec.ProbeTimeout) * time.Second
+	}
+
+	resourceRefs := []runtimev1alpha1.TypedReference{}
+	for _, workloadRef := range healthScope.Status.WorkloadReferences {
+		// Get workload object.
+		workloadObject := unstructured.Unstructured{}
+		workloadObject.SetAPIVersion(workloadRef.APIVersion)
+		workloadObject.SetKind(workloadRef.Kind)
+		workloadObjectRef := types.NamespacedName{Namespace: healthScope.GetNamespace(), Name: workloadRef.Name}
+		if err := client.Get(ctx, workloadObjectRef, &workloadObject); err != nil {
+			return errors.Wrapf(err, errNoWorkload, workloadRef.Name)
+		}
+
+		// TODO(asouza): Need to find a way to make sure that this is compatible with the scope structure.
+		if value, err := fieldpath.Pave(workloadObject.UnstructuredContent()).GetValue("status.resources"); err == nil {
```

Comment:
why does the workload status have a "resources" field?

Reply:
It shows which resources were generated for a given workload. This is how I find which resources to do the health check.

Reply:
not all workload have the "resource" field, as a matter of fact, most don't